src/FreeRTOS/heap_4_infinitime.c

```diff
@@ -0,0 +1,517 @@
+ pxEnd = ( void * ) uxAddress;
+ pxEnd->xBlockSize = 0;
+ pxEnd->pxNextFreeBlock = NULL;
+
+ /* To start with there is a single free block that is sized to take up the
+ entire heap space, minus the space taken by pxEnd. */
+ pxFirstFreeBlock = ( void * ) pucAlignedHeap;
+ pxFirstFreeBlock->xBlockSize = uxAddress - ( size_t ) pxFirstFreeBlock;
+ pxFirstFreeBlock->pxNextFreeBlock = pxEnd;
+
+ /* Only one block exists - and it covers the entire usable heap space. */
+ xMinimumEverFreeBytesRemaining = pxFirstFreeBlock->xBlockSize;
+ xFreeBytesRemaining = pxFirstFreeBlock->xBlockSize;
+
+ /* Work out the position of the top bit in a size_t variable. */
+ xBlockAllocatedBit = ( ( size_t ) 1 ) << ( ( sizeof( size_t ) * heapBITS_PER_BYTE ) - 1 );
+}
+/*-----------------------------------------------------------*/
+
+static void prvInsertBlockIntoFreeList( BlockLink_t *pxBlockToInsert )
+{
+ BlockLink_t *pxIterator;
+ uint8_t *puc;
+
+ /* Iterate through the list until a block is found that has a higher address
+ than the block being inserted. */
+ for( pxIterator = &xStart; pxIterator->pxNextFreeBlock < pxBlockToInsert; pxIterator = pxIterator->pxNextFreeBlock )
+ {
+   /* Nothing to do here, just iterate to the right position. */
+ }
+
+ /* Do the block being inserted, and the block it is being inserted after
+ make a contiguous block of memory? */
+ puc = ( uint8_t * ) pxIterator;
+ if( ( puc + pxIterator->xBlockSize ) == ( uint8_t * ) pxBlockToInsert )
+ {
+   pxIterator->xBlockSize += pxBlockToInsert->xBlockSize;
+   pxBlockToInsert = pxIterator;
+ }
+ else
+ {
+   mtCOVERAGE_TEST_MARKER();
+ }
+
+ /* Do the block being inserted, and the block it is being inserted before
+ make a contiguous block of memory? */
+ puc = ( uint8_t * ) pxBlockToInsert;
+ if( ( puc + pxBlockToInsert->xBlockSize ) == ( uint8_t * ) pxIterator->pxNextFreeBlock )
+ {
+   if( pxIterator->pxNextFreeBlock != pxEnd )
+   {
+     /* Form one big block from the two blocks. */
+     pxBlockToInsert->xBlockSize += pxIterator->pxNextFreeBlock->xBlockSize;
+     pxBlockToInsert->pxNextFreeBlock = pxIterator->pxNextFreeBlock->pxNextFreeBlock;
+   }
+   else
+   {
+     pxBlockToInsert->pxNextFreeBlock = pxEnd;
+   }
+ }
+ else
+ {
+   pxBlockToInsert->pxNextFreeBlock = pxIterator->pxNextFreeBlock;
+ }
+
+ /* If the block being inserted plugged a gab, so was merged with the block
+ before and the block after, then it's pxNextFreeBlock pointer will have
+ already been set, and should not be set here as that would make it point
+ to itself. */
+ if( pxIterator != pxBlockToInsert )
+ {
+   pxIterator->pxNextFreeBlock = pxBlockToInsert;
+ }
+ else
+ {
+   mtCOVERAGE_TEST_MARKER();
+ }
+}
+
+/*-----------------------------------------------------------*/
+
+void *pvPortRealloc( void *pv, size_t xWantedSize )
+{
+ size_t move_size;
+ size_t block_size;
+ BlockLink_t* pxLink;
+ void* pvReturn = NULL;
+ uint8_t* puc = (uint8_t*) pv;
+
+ if (xWantedSize > 0)
+ {
+   if (pv != NULL)
+   {
```

Comment:
I think it would make this easier to read if these checks were reversed and used as guard clauses.

Reply:
I agree. However, I tried to follow the same coding style than the other functions from FreeRTOS. They all have a single `return` statement (I mean at least in this file).

If we decide to take a bit of freedom in the coding style, here's how it could look : 

```
void *pvPortRealloc( void *pv, size_t xWantedSize )
{
 size_t move_size;
 size_t block_size;
 BlockLink_t* pxLink;
 void* pvReturn = NULL;
 uint8_t* puc = (uint8_t*) pv;

 // Zero bytes requested, do nothing (according to libc, this behavior implementation defined)
 if(xWantedSize == 0) {
   return NULL;
 }

 if( pv == NULL) {
   // pv points to NULL. Allocate a new buffer.
   return pvPortMalloc(xWantedSize);
 }

 // The memory being freed will have an BlockLink_t structure immediately before it.
 puc -= xHeapStructSize;

 // This casting is to keep the compiler from issuing warnings.
 pxLink = (void*) puc;

 // Check allocate block
 if ((pxLink->xBlockSize & xBlockAllocatedBit) != 0)
 {
   // The block is being returned to the heap - it is no longer allocated.
   block_size = (pxLink->xBlockSize & ~xBlockAllocatedBit) - xHeapStructSize;

   // Allocate a new buffer
   pvReturn = pvPortMalloc(xWantedSize);

   // Check creation and determine the data size to be copied to the new buffer
   if (pvReturn != NULL)
   {
     if (block_size < xWantedSize)
     {
       move_size = block_size;
     }
     else
     {
       move_size = xWantedSize;
     }

     // Copy the data from the old buffer to the new one
     memcpy(pvReturn, pv, move_size);

     // Free the old buffer
     vPortFree(pv);
   }
 }
 else
 {
   // pv does not point to a valid memory buffer. Allocate a new one
   pvReturn = pvPortMalloc(xWantedSize);
 }

  return pvReturn;
}
```

What do you think?

Reply:
That looks like what I had in mind. I don't think that trying to stick to FreeRTOS's code style is necessary unless we are intending to upstream it.